bluebutton-data-pipeline-rif-extract/src/main/java/gov/hhs/cms/bluebutton/datapipeline/rif/extract/s3/DataSetMonitorWorker.java

```diff
@@ -0,0 +1,283 @@
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.DeleteObjectsRequest;
+import com.amazonaws.services.s3.model.DeleteObjectsRequest.KeyVersion;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.ListObjectsRequest;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+
+import gov.hhs.cms.bluebutton.datapipeline.rif.extract.s3.DataSetManifest.DataSetManifestEntry;
+import gov.hhs.cms.bluebutton.datapipeline.rif.model.RifFile;
+import gov.hhs.cms.bluebutton.datapipeline.rif.model.RifFilesEvent;
+
+/**
+ * <p>
+ * Acts as a worker {@link Runnable} for {@link DataSetMonitor}. It is expected that
+ * this will be run on a repeating basis, via a {@link ScheduledExecutorService}
+ * .
+ * </p>
+ * <p>
+ * When executed via {@link #run()}, the {@link DataSetMonitorWorker} will scan the
+ * specified Amazon S3 bucket. It will look for <code>manifest.xml</code>
+ * objects/files and select the oldest one available. If such a manifest is
+ * found, it will then wait for all of the objects in the data set represented
+ * by it to become available. Once they're all available, it will kick off the
+ * processing of the data set, and block until that processing has completed.
+ * </p>
+ */
+final class DataSetMonitorWorker implements Runnable {
+	private static final Pattern REGEX_MANIFEST = Pattern.compile("^(.*)\\/manifest\\.xml$");
+
+	private final String bucketName;
+	private final DataSetProcessor dataSetProcessor;
+	private final AmazonS3 s3Client = new AmazonS3Client(new DefaultAWSCredentialsProviderChain());
+
+	/**
+	 * Constructs a new {@link DataSetMonitorWorker} instance.
+	 * 
+	 * @param bucketName
+	 *            the name of the AWS S3 bucket to monitor
+	 * @param dataSetProcessor
+	 *            the {@link DataSetProcessor} responsible for processing each
+	 *            of the {@link RifFilesEvent}s that are generated by this
+	 *            {@link DataSetMonitorWorker}
+	 */
+	public DataSetMonitorWorker(String bucketName, DataSetProcessor dataSetProcessor) {
+		this.bucketName = bucketName;
+		this.dataSetProcessor = dataSetProcessor;
+	}
+
+	/**
+	 * @see java.lang.Runnable#run()
+	 */
+	@Override
+	public void run() {
+		/*
+		 * Request a list of all objects in the configured bucket and directory.
+		 * (In the results, we'll be looking for the oldest manifest file, if
+		 * any.)
+		 */
+		ListObjectsRequest bucketListRequest = new ListObjectsRequest();
+		bucketListRequest.setBucketName(bucketName);
+
+		/*
+		 * S3 will return results in separate pages. Loop through all of the
+		 * pages, looking for the oldest manifest file that is available.
+		 */
+		String manifestToProcessKey = null;
+		ObjectListing objectListing = s3Client.listObjects(bucketListRequest);
```

Comment:
> Anything that gets manually modified like this should probably be moved to a separate bucket.

I agree! But "the ETL service errored out because I renamed a directory" seems like surprising behavior for sysadmins/ops folk. I'd rather leave this in. 

> There is more danger, I think, in a scenario where...

I hadn't considered that scenario, so thanks for pointing it out! But what about when bucket #1 just has a single RIF file, and both it and the manifest fail to transfer due to a network outage? We also need to catch those problems: no objects have the missing keys.

I think there are two possible correct solutions here:

1. Catcher side: Assign every data set an index, which must be a monotonic, consecutive integer? That way if the data set IDs go `1, 2, 4`, we know we're missing something?
2. Pitcher side: Ensure that the transfer process guards against pushing data sets out of order, perhaps with a lock file and verifying that objects are present in S3 before removing it. Or something like that.

I agree that we need to address this, but I'm going to file a separate issue for it, to allow us time to discuss with @skoerselman.